accord-core/src/main/java/accord/local/Commands.java

```diff
@@ -1006,13 +1026,14 @@ public void accept(SafeCommandStore safeStore)
                         {
                             default: throw new AssertionError("Unexpected redundant status: " + redundantStatus);
                             case NOT_OWNED: throw new AssertionError("Invalid state: waiting for execution of command that is not owned at the execution time");
-                            case LOCALLY_REDUNDANT:
-                            case PRE_BOOTSTRAP:
-                                removeRedundantDependencies(safeStore, prevSafe, txnIds[depth], redundantStatus == PRE_BOOTSTRAP);
+                            case SHARD_REDUNDANT:
+                            case PARTIALLY_REDUNDANT_PRE_BOOTSTRAP_OR_STALE:
+                            case PRE_BOOTSTRAP_OR_STALE:
+                                removeRedundantDependencies(safeStore, prevSafe, txnIds[depth], redundantStatus == PRE_BOOTSTRAP_OR_STALE);
```

Comment:
`appliedOrInvalidated` is a collection we maintain for the benefit of transactions that depend on us; they will explore their dependencies from most recent backwards, reading the appliedOrInvalidated bitsets and propagating that information to their own dependency bit sets, saving them reading the state of any coinciding dependencies.

Reply:
Regarding redundancy, staleness is managed by `RedundantBefore` and is a form of redundancy IMO - we no longer consider the transaction relevant to the execution of other transactions since we're stale and will either be re-bootstrapping (to be implemented) or have simply lost the range a little early.

Reply:
`removeRedundantDependencies` is invoked when we encounter a dependency that is redundant (or stale, or pre bootstrap or whatever), i.e. that we expect _should_ have been possible to infer the redundancy of via `RedundantBefore`. So we go and re-apply any `RedundantBefore` knowledge - that is presumably more up to date than it was when we first constructed `WaitingOn`. We then make sure the transaction in question is removed from the `WaitingOn` set, and if it is safe to do so we set the associated propagating bit in `WaitingOn` so that later transactions may inherit our deduction.